agent/tool-scripts/datalog/prometheus-metrics-datalog

```diff
@@ -2,33 +2,69 @@
-import os,subprocess,warnings,logging
+import os, subprocess, warnings, logging
 
 interval = int(sys.argv[2])
 host = sys.argv[1]
 tool_log = sys.argv[3]
 port = int(sys.argv[4])
 cert = sys.argv[5]
-key = sys.argv[6]
+selfsigned = sys.argv[6]
+key = sys.argv[7]
+
+if (cert == "None") or (key == "None"):
+    command = "prom2json http://%s:%d/metrics" % (host, port)
+else:
+    command = "prom2json %s --cert=%s --key=%s https://%s:%d/metrics" % (
+            "--accept-invalid-cert" if selfsigned == "true" else "", cert, key, host, port)
+
+# Initial timestamp and next interval calcultion
+time_stamp = int(time.time())
+next_start = time_stamp + interval
+intervals = 0
+total_duration = 0
+drifted = 0
+
 while True:
-    time_stamp = int(time.time())
-    next_start = time_stamp + interval
-    if (cert == "None") or (key == "None"):
-        command = "prom2json http://%s:%d/metrics" % (host, port)
-    else:
-        command = "prom2json --cert=%s --key=%s https://%s:%d/metrics" % (cert, key, host, port)
     command_out = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE)
     prom2json_out = command_out.communicate()[0]
-    metrics_list = json.loads(prom2json_out)
-    if type(metrics_list) is list:
-        for item in metrics_list:
-            del item['help']
-            item["@timestamp"] = time_stamp
-            print json.dumps(item)
-        endtime = time.time()
-        if endtime > next_start:
-            logging.warn("[%d] [prometheus-metrics] please change the interval to atleast %d seconds\n" %(time_stamp, endtime - next_start))
-        if endtime < next_start:
-            time.sleep(next_start - endtime)
+
+    # Record how long prom2json took for use when we run beyond the
+    # given interval.
+    end_cmd_time_stamp = int(time.time())
+
+    try:
+        metrics_list = json.loads(prom2json_out)
+    except Exception as e:
+        logging.warn("[prometheus-metrics-datalog] json.loads(prom2json_out) failed with %r" % e)
     else:
-        logging.warn("json.load(prom2json_out) did not return a list")
+        if type(metrics_list) is list:
+            for item in metrics_list:
+                del item['help']
+                item["@timestamp"] = time_stamp
+                print json.dumps(item)
+        else:
+            logging.warn("[prometheus-metrics-datalog] json.load(prom2json_out) did not return a list")
+
+    # Record how long all the work took during this interval.
+    endtime = time.time()
+
+    duration = endtime - time_stamp
+    total_duration += duration
+    avg_duration = total_duration / intervals
+
+    if endtime < next_start:
+        # Sleep for the remainder of the interval
+        time.sleep(next_start - endtime)
+    elif endtime > next_start:
+        drifted += 1
+        logging.warn("[prometheus-metrics-datalog] interval exceeded now %d times: [interval: %d, start: %d, duration: %d, prom2json: %d]; please consider changing the interval to at least %d seconds\n" % (
```

Comment:
This logging line is broken, not sure why. Every time it's uncommented the line that follows this throws a syntax error.

Reply:
Yeah, missing parenthesis at the end of the line.  My bad.

I posted a new PR, https://github.com/sjug/pbench/pull/2, after I wrote a quick test program to verify.  Shame on me for not doing that earlier.